examples/doremi/doremi/dataloader.py

```diff
@@ -0,0 +1,363 @@
+        if current_pp_rank == self.output_pp_rank:
+            result["label_ids"] = input_ids[:, 1:]
+            result["label_mask"] = np.ones((batch_size, self.sequence_length), dtype=np.bool_)
+
+        result["domain_idxs"] = np.vstack([examples[i]["domain_ids"] for i in range(len(examples))])
+
+        if isinstance(result["input_ids"], torch.Tensor) and result["input_ids"].shape[-1] != self.sequence_length:
+            raise ValueError(
+                f"`labels` are incorrectly preprocessed. `labels` length is {result['input_ids'].shape[-1]}, but should be"
+                f" {self.sequence_length}."
+            )
+        if isinstance(result["label_ids"], torch.Tensor) and result["label_ids"].shape[-1] != self.sequence_length:
+            raise ValueError(
+                f"`labels` are incorrectly preprocessed. `labels` length is {result['label_ids'].shape[-1]}, but should be"
+                f" {self.sequence_length}."
+            )
+
+        # Cast np.array to torch.Tensor
+        result = {k: v if isinstance(v, TensorPointer) else torch.from_numpy(v) for k, v in result.items()}
+        return result
+
+
+class DistributedSamplerForDoReMi(DistributedSampler):
+    def __init__(
+        self,
+        datasets: List[Dataset],
+        batch_size: int,
+        num_microbatches: int,
+        shuffle: bool = False,
+        seed: int = 42,
+        doremi_context: Optional[DoReMiContext] = None,
+        parallel_context: Optional[ParallelContext] = None,
+        **kwargs,
+    ):
+        assert len(datasets) == len(
+            doremi_context.domain_weights
+        ), "The number of datasets must equal to the number of domain weights"
+        assert doremi_context is not None
+        assert parallel_context is not None
+
+        super().__init__(datasets, **kwargs)
+
+        self.datasets = datasets
+        self.batch_size = batch_size
+        self.num_microbatches = num_microbatches
+        self.shuffle = shuffle
+        self.doremi_context = doremi_context
+        self.parallel_context = parallel_context
+        self.total_size = self._calculate_total_size()
+
+        self.lengths = [len(d) for d in self.datasets]
+        self.offsets = np.cumsum([0] + self.lengths[:-1])
+        self.seed = seed
+
+        dp_size = dist.get_world_size(self.parallel_context.dp_pg)
+        self.global_batch_size = batch_size * dp_size * num_microbatches
+        # NOTE: Reset the seed of the generator for consistent randomness across epochs
+        self.generator = torch.Generator(device="cpu").manual_seed(
+            seed * (1 + dist.get_rank(self.parallel_context.dp_pg)) * (1 + dist.get_rank(self.parallel_context.pp_pg))
+        )
+
+        self.reset()
+
+    def _calculate_total_size(self):
+        total_samples = sum(len(d) for d in self.datasets)
+        return math.ceil(total_samples / self.batch_size) * self.batch_size
+
+    def __iter__(self):
+        return self
```

Comment:
In the current implementation, the state of an iterator of a class is stored in its instance's attributes. So, just in case a user runs iter(DistributedSamplerForDoReMi()), we return the instance itself